README.md
# ReCalc v1.3.4

Calculadora con funciones simples.

## Dependencias

- node > 18

## Comandos

```bash
# iniciar/limpiar la db
npm run seed

# iniciar calculadora
npm run start

# correr lost test
npm run test
```

## Lista de integrantes del grupo

- Gomez D'Addario, Agustin
- de la Mano, Manuel Agustín
- Mijael Ben, Ulises Jose
- Lorenzo, Nahuel Emiliano

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.3.4] 19-07-2023

### Added

- Se agregó test e2e para cuando se realiza una potencia mayor a 100000
- Se agregó test e2e para la división por cero

## [1.3.3] 20-06-2023

### Added 

- Se agregó botón de pow
- Se agregó función y botón de binary
- Se realizó conexiones del front con la API
- Se agregó el botón de raiz
- Se agregaron test e2e

### Fixed

- Se solucionaron los errores de los test estáticos
- Se solucionó bug del botón igual
- Se solucionó bug del botón C
- Se solucionó bug para restar números negativos

## [1.3.2] 08-06-2023

### Fixed

- Path donde corre eslint

### Fixed

- Problema al correr tests e2e en Windows

## [1.3.1] 05-06-2023

### Added

- Dependencia cross-env para correr tests e2e

### Fixed

- Problema al correr tests e2e en Windows

## [1.3.0] 05-06-2023

### Added

- Interfaz básica de la calculadora
- ESLint para tests estáticos
- Playwright para tests e2e

## [1.2.0] 15-05-2023

### Added

- Sequelize para el manejo de base de datos
- Tests de integración sobre API y modelo
- Integración continua

## [1.1.0] 20-04-2023

### Added

- Endpoint para la función de resta
- Dependencias para realizar tests
- Tests para la función de resta

## [1.0.0] 17-04-2023

### Added

- Interface CLI para realizar cálculos en forma interactiva
- Base de la API
- Funciones core para realizar suma, resta, multiplicación, división y potencia

[unreleased]: https://github.com/frlp-utn-ingsoft/recalc/compare/v1.3.4...HEAD
[1.3.4]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.4
[1.3.3]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.3
[1.3.2]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.2
[1.3.1]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.1
[1.3.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.0
[1.2.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.2.0
[1.1.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.1.0
[1.0.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.0.0

test/ui.test.js
import { test, expect } from '@playwright/test';
import { seed } from '../src/seed.js';
import { Operation, History } from '../src/models.js'
import { error } from 'console';

test.describe('test', () => {
  test.describe.configure({ mode: 'serial' });

  test.beforeEach(async () => {
    await seed();
  })

  test('Deberia tener como titulo de pagina recalc', async ({ page }) => {
    await page.goto('./');

    // Expect a title "to contain" a substring.
    await expect(page).toHaveTitle(/recalc/i);
  });

  test('Deberia poder realizar una resta', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '7' }).click()
    await page.getByRole('button', { name: '9' }).click()
    await page.getByRole('button', { name: '-' }).click()
    await page.getByRole('button', { name: '9' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/sub/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(70);

    await expect(page.getByTestId('display')).toHaveValue(/70/)

    const operation = await Operation.findOne({
      where: {
        name: "SUB"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(79)
    expect(historyEntry.secondArg).toEqual(9)
    expect(historyEntry.result).toEqual(70)
  });

  test('Deberia poder realizar una suma', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '7' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '+' }).click()
    await page.getByRole('button', { name: '9' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/add/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(79);

    await expect(page.getByTestId('display')).toHaveValue(/79/)

    const operation = await Operation.findOne({
      where: {
        name: "ADD"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(70)
    expect(historyEntry.secondArg).toEqual(9)
    expect(historyEntry.result).toEqual(79)
  });

  test('Deberia poder realizar una multiplicacion', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '1' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '*' }).click()
    await page.getByRole('button', { name: '5' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/mul/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(50);

    await expect(page.getByTestId('display')).toHaveValue(/50/)

    const operation = await Operation.findOne({
      where: {
        name: "MUL"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(10)
    expect(historyEntry.secondArg).toEqual(5)
    expect(historyEntry.result).toEqual(50)
  });  

   test('Deberia poder realizar una potencia', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '7' }).click()
    await page.getByRole('button', { name: '^2' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/pow/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(49);

    await expect(page.getByTestId('display')).toHaveValue(/49/)

    const operation = await Operation.findOne({
      where: {
        name: "POW"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })
    expect(historyEntry.firstArg).toEqual(7)
    expect(historyEntry.result).toEqual(49)
  });

  test('Deberia poder realizar una división', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '9' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '/' }).click()
    await page.getByRole('button', { name: '9' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/div/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(10);

    await expect(page.getByTestId('display')).toHaveValue(/10/)

    const operation = await Operation.findOne({
      where: {
        name: "DIV"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(90)
    expect(historyEntry.secondArg).toEqual(9)
    expect(historyEntry.result).toEqual(10)
  });

  test('Deberia poder realizar la raiz cuadrada', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '9' }).click()
    await page.getByRole('button', { name: 'square' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/sqrt/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(3);

    await expect(page.getByTestId('display')).toHaveValue(/3/)

    const operation = await Operation.findOne({
      where: {
        name: "SQRT"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(9)
    expect(historyEntry.result).toEqual(3)
  });

  test('Deberia poder ingresar un numero dar click en igual y que no devuelva undefined', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '8' }).click()
    await page.getByRole('button', { name: '=' }).click()

    await expect(page.getByTestId('display')).toHaveValue(/8/)

  });
  
    test('Debria poder borrar una cuenta de la interfaz con C', async ({ page }) => {
      await page.goto('./');
  
      await page.getByRole('button', { name: '9' }).click()
      await page.getByRole('button', { name: '0' }).click()
      await page.getByRole('button', { name: '/' }).click()
      await page.getByRole('button', { name: 'c' }).click()
  
      await expect(page.getByTestId('display')).toHaveValue(/0/)
    });
  
    test('Deberia poder realizar una conversion de decimal a binario', async ({ page }) => {
      await page.goto('./');
  
      await page.getByRole('button', { name: '9' }).click()
      await page.getByRole('button', { name: 'binary' }).click()
  
      const [response] = await Promise.all([
        page.waitForResponse((r) => r.url().includes('/api/v1/binary/')),
        page.getByRole('button', { name: '=' }).click()
      ]);
  
      const { result } = await response.json();
      expect(result).toBe(1001);
  
      await expect(page.getByTestId('display')).toHaveValue(/1001/)
  
      const operation = await Operation.findOne({
        where: {
          name: "BINARY"
        }
      });
  
      const historyEntry = await History.findOne({
        where: { OperationId: operation.id }
      })
  
      expect(historyEntry.firstArg).toEqual(9)
      expect(historyEntry.result).toEqual(1001)
    });

    test('Deberia realizar una resta con numeros negativos', async ({ page }) => {
      await page.goto('./');
    
      await page.getByRole('button', { name: '-' }).click()
      await page.getByRole('button', { name: '6' }).click()
      await page.getByRole('button', { name: '3' }).click()
      await page.getByRole('button', { name: '-' }).click()
      await page.getByRole('button', { name: '3' }).click()
    
      const [response] = await Promise.all([
        page.waitForResponse((r) => r.url().includes('/api/v1/sub/')),
        page.getByRole('button', { name: '=' }).click()
      ]);
    
      const { result } = await response.json();
      expect(result).toBe(-66);
    
      await expect(page.getByTestId('display')).toHaveValue(/-66/)
    
      const operation = await Operation.findOne({
        where: {
          name: "SUB"
        }
      });
    
      const historyEntry = await History.findOne({
        where: { OperationId: operation.id }
      })
    
      expect(historyEntry.firstArg).toEqual(-63)
      expect(historyEntry.secondArg).toEqual(3)
      expect(historyEntry.result).toEqual(-66)
    });

    test('Deberia lanzar un error cuando se quiere realizar una potencia con un numero mayor a 100000', async ({ page }) => {
      await page.goto('./');
  
      await page.getByRole('button', { name: '9' }).click()
      await page.getByRole('button', { name: '9' }).click()
      await page.getByRole('button', { name: '9' }).click()
      await page.getByRole('button', { name: '9' }).click()
      await page.getByRole('button', { name: '9' }).click()
      await page.getByRole('button', { name: '9' }).click()
      await page.getByRole('button', { name: '^2' }).click()
  
        page.getByRole('button', { name: '=' }).click()
  
        await expect(page.getByTestId('display')).toHaveValue("error")

    });
    
    test('Deberia lanzar un error cuando se quiere realizar una division por cero', async ({ page }) => {
      await page.goto('./');

      await page.getByRole('button', { name: '5' }).click()
      await page.getByRole('button', { name: '/' }).click()
      await page.getByRole('button', { name: '0' }).click()

        page.getByRole('button', { name: '=' }).click()

        await expect(page.getByTestId('display')).toHaveValue("Math error")

    });
})

test/api.test.js
const request = require('supertest');
const api = require('../src/api.js');
const { seed } = require('../src/seed.js');
const { getHistory } = require('../src/models.js');
const { createHistoryEntry } = require('../src/models.js');

beforeEach(async () => {
    await seed()
})

describe("API substract", () => {
    test("Deberia responder con un 200 ok", async () => {
        const app = await api.build()

        return request(app).get('/api/v1/sub/2/1')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toEqual(1);
            })
    })
})

describe("API multiply", () => {
    test("Deberia responder con un 200 ok", async () => {
        const app = await api.build()

        return request(app).get('/api/v1/mul/3.1/5.4')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).not.toEqual(parseInt(3.1*5.4));
            })
    })
})

describe("API add", () => {
    test("Si segundo parámetro es negativo, el resultado tiene que ser menor al primer parámetro y el endpoint devuelver un status 200.", async () => {
        const app = await api.build()

        return request(app).get('/api/v1/add/2/-1')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toBeLessThan(2);
            })
    })
})

describe("API add", () => {
    test("La suma de 0.1 con 0.2, debe dar 0.3.", async () => {
        const app = await api.build()
        
        return request(app).get('/api/v1/add/0.1/0.2')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")

            .then((res) => {
                expect(res.body.result).toBeCloseTo(0.3);
            })
    })
})

describe("API power", () => {
    test("Deberia responder con un 400 Error", async () => {
        const app = await api.build()

        return request(app).get('/api/v1/pow/a')
            .expect(400)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((err) => {
                expect(err.body.error).toEqual("Uno de los parámetros no es un número"); 
            })
    })
})


describe("API id", () => {
    test("Deberia responder con un 404 ok", async () => {
        const app = await api.build()

        return request(app).get('/api/v1/id/45')
            .expect(404)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((err) => {
                expect(err.body.error).toEqual("No se encontró ninguna entrada de historial con el ID proporcionado");
            })      
    })
})

describe("API div", () => {
    test("si el segundo parámetro es 0, el endpoint devuelve el mensaje de error con un status 400.", async () => {
        const app = await api.build()

        return request(app).get('/api/v1/div/2/0')
            .expect(400)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.error).toEqual("El divisor no puede ser cero");
            })
    })
})

describe("API sqrt", () => {
    test("La raiz cuadrada de 100, debe dar 10.", async () => {
        const app = await api.build()
        
        return request(app).get('/api/v1/sqrt/100')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")

            .then((res) => {
                expect(res.body.result).toEqual(10);
            })
    })
})

describe("History deleted All", () => {
    test("Deberia eliminar History con un status 200.", async () => {
        const app = await api.build()
        await createHistoryEntry({
            firstArg: 2,
            secondArg: 2,
            result: 0,
            operationName: "SUB"
        });
        await createHistoryEntry({
            firstArg: 2,
            secondArg: 2,
            result: 0,
            operationName: "SUB"
        })
        const cantidad = (await getHistory()).length
        return request(app).get('/api/v1/allDeleted')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toEqual(cantidad + " registros eliminados de la tabla History.");
            })
    })
})

describe("History All", () => {
    test("Deberia retornar todo History con un status 200.", async () => {
        const app = await api.build()
        const cantidad = await getHistory()
        return request(app).get('/api/v1/all')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toEqual(cantidad);
            })
    })
})

describe("API Binary", () => {
    test("El numero binario de 999 deberia ser 1111100111", async () => {
        const app = await api.build()
        
        return request(app).get('/api/v1/binary/999')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")

            .then((res) => {
                expect(res.body.result).toEqual(1111100111);
            })
    })
})

test/core.test.js
import core from '../src/core.js'

describe('Subtract', () => {
    test('Deberia 2 - 2 = 0', () => {
        expect(core.sub(2, 2)).toBe(0); 
    })

    test('Deberia 6 - 4 = 2', () => {
        expect(core.sub(6, 4)).toBe(2); 
    })

    test('Deberia 6 - 8 = -2', () => {
        expect(core.sub(6, 8)).toBeLessThan(0); 
    })
})

describe('Multiply', () => {
    test('Deberia 3 * 3 = 9', () => {
    expect(core.mul(3, 3)).toBe(9); 
    })

    test('Deberia 8 * -6 < 0', () => {
        expect(core.mul(8, -6)).toBeLessThan(0); 
    })
})

describe('Add', () => {
    test('Deberia 2 + 5 = 7', () => {
        expect(core.add(2, 5)).toBe(7);
    })

    test('Deberia -5 + 5', () => {
        expect(core.add(-5, 5)).toBe(0)
    })
})


describe('divide', () => {
    test('Deberia 2 / 2 = 1', () => {
        expect(core.div(2, 2)).toBe(1); 
    })

    test('Deberia 12 / 0 = Math Error', () => {
        expect(core.div(12, 0)).toBe("Math Error"); 
    })
})


describe('Pow', () => {
    test('Deberia 0 ^ 2 = 0', () => {
        expect(core.pow(0)).toBe(0); 
    })

    test('Deberia 12 ^ 2 = 144', () => {
        expect(core.pow(12)).toBe(144); 
    })

    test('Deberia -2 ^ 2 = numero positivo', () => {
        expect(core.pow(-2)).toBeGreaterThan(0);
    })
})

describe('Binary', () => {
    test('Deberia binary 0 = 0', () => {
        expect(core.binary(0)).toBe(0); 
    })

    test('Deberia binary 23 = 10111', () => {
        expect(core.binary(23)).toBe(10111); 
    })

    test('Deberia binary 10 = 1010', () => {
        expect(core.binary(10)).toBe(1010);
    })
})

src/cli.js
import { createInterface } from "node:readline/promises";
import core from "./core.js"

const readline = createInterface({
    input: process.stdin,
    output: process.stdout,
});

const AVAILABLE_FNS = [...Object.keys(core), 'exit'].join(', ')

async function loop() {
    const fnName = await readline.question(`Ingrese función (${AVAILABLE_FNS}): `)
 
    if (fnName === "exit") {
        console.log("👋👋👋");
        return readline.close();
    }
    else if(fnName!=="add" && fnName!=="sub" && fnName!=="mul" && fnName!=="div" && fnName!=="pow" && fnName!=="sqrt" && fnName!=="binary") {
        console.log("La funcion ingresada no esta disponible");
        loop();
    }

    const fn = core[fnName];

    let result
    if (fnName === "pow" || fnName === "sqrt" || fnName === "binary") {
        const num = await readline.question("Ingrese el número: ")
        result = fn(Number(num))
    }
    else {
        const firstNum = await readline.question("Ingrese el primer número: ")
        const secondNum = await readline.question("Ingrese el primer número: ")
        result = fn(Number(firstNum), Number(secondNum));
    }

    console.log(result)
    loop();
}

loop();


src/models.js
import { Sequelize, DataTypes } from 'sequelize';

const inTest = process.env.NODE_ENV === 'test';

const sequelize = new Sequelize({
    dialect: 'sqlite',
    logging: !inTest,
    storage: inTest ? './db.sqlite3' : './db.sqlite3'
})

export const History = sequelize.define('History', {
    firstArg: {
        type: DataTypes.NUMBER,
        allowNull: false
    },
    secondArg: {
        type: DataTypes.NUMBER,
        allowNull: true
    },
    result: {
        type: DataTypes.NUMBER,
        allowNull: true
    },
    error: {
        type: DataTypes.STRING,
        allowNull: true
    }
});

export const Operation = sequelize.define('Operation', {
    name: {
        type: DataTypes.STRING,
        allowNull: false
    },
});

Operation.hasMany(History)
History.belongsTo(Operation)

export async function createHistoryEntry({ firstArg, secondArg, operationName, result, error }) {
    const operation = await Operation.findOne({
        where: {
            name: operationName
        }
    });

    return History.create({
        firstArg,
        secondArg,
        result,
        OperationId: operation.id,
        error
    })
}

export function createTables() {
    return Promise.all([
        History.sync({ force: true }),
        Operation.sync({ force: true })
    ]);
}

export async function getHistory(operationName = null) {
    try {
        let res
        if (operationName) {
            res = await History.findAll({
                include: Operation,
                where: {
                    "$Operation.name$": operationName
                }
            });
        } else {
            res =  await History.findAll({ include: Operation });
        }
        return res;
    } catch (error) {
        throw new Error('Error al obtener el historial de operaciones: ' + error.message);
    }
}

export function deleteAllHistory() {
    return History.destroy({
        where: {}
    })
}

export function getHistoryEntryById(id) {
    return History.findByPk(id, {
        include: Operation
    });
}

src/public/index.js
const $display = document.querySelector('.display')
const $buttons = document.querySelector('.buttons')

const operations = ['-','*','+', '^2','/', 'square', 'binary'];

let currentDisplay = "";
let operation = null;
let reset = false;


$buttons.addEventListener('click', async (e) => {
    const nextAction = e.target.name

    if (nextAction === "=") {
        let negative = false;
        if (currentDisplay.startsWith("-")&& operation === "-")
        {
            currentDisplay = currentDisplay.slice(1)
            negative = true;
        }
        const [firstArg, secondArg] = currentDisplay.split(operation)

        let result;
        
        if (operation === "-") {
            result = await calculateSub(negative ? firstArg*(-1): firstArg, secondArg)
        }

        if (operation === "binary") {
            result = await calculateBinary(firstArg)
        }

        if (operation === "^2"){
            if (parseInt(firstArg) > 100000){
                result = "error"
            }
            else{result = await calculatePow(firstArg)}
        } 

        if (operation === "/") {
            if(secondArg === "0"){
                result = "Math error"
            }
            else{
                result = await calculateDiv(firstArg, secondArg)
            }
        }
        
        if (operation === "*") {
            result = await calculateMul(firstArg, secondArg)
        }

        if (operation === "+") {
            result = await calculateAdd(firstArg, secondArg)
        }
    
        if (operation === "square"){
            result = await calculateSqrt(firstArg)
        }

        reset = true;
        if (result !== undefined) {
            result = renderDisplay(result);
        } else {
            result = renderDisplay(currentDisplay);
        }
        return result;
    }

    if (operations.includes(nextAction)) {
        operation = nextAction;
    }

    if (nextAction === "c") {
        return renderDisplay(0);
    }
    
    if (reset || currentDisplay === 0) {
        reset = false;
        operation = null;
        renderDisplay(nextAction);
    } else {
        renderDisplay(currentDisplay + nextAction);
    }
})

async function calculateSub(firstArg, secondArg) {
    const resp = await fetch(`/api/v1/sub/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculatePow(firstArg) {
    const resp = await fetch(`/api/v1/pow/${firstArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateDiv(firstArg, secondArg) {
    const resp = await fetch(`/api/v1/div/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateMul(firstArg, secondArg) {
    const resp = await fetch(`/api/v1/mul/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateAdd(firstArg, secondArg) {
    const resp = await fetch(`/api/v1/add/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateSqrt(firstArg) {
    const resp = await fetch(`/api/v1/sqrt/${firstArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateBinary(firstArg) {
    const resp = await fetch(`/api/v1/binary/${firstArg}`)
    const { result } = await resp.json();

    return result;
}

function renderDisplay(chars) {
    currentDisplay = chars;
    $display.value = chars;
}

src/controllers.js
import express from 'express';
import core from './core.js';

import { createHistoryEntry, deleteAllHistory, getHistory, getHistoryEntryById } from "./models.js";

const router = express.Router();

router.get("/sub/:a/:b", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);
    const b = Number(params.b);

    if (isNaN(a) || isNaN(b)) {
        res.status(400).send('Uno de los parámetros no es un número');
    } else {
        const result = core.sub(a, b);

        await createHistoryEntry({ firstArg: a, secondArg: b, result, operationName: "SUB", error: null })
        return res.send({ result });
    }
});

router.get("/add/:a/:b", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);
    const b = Number(params.b);

    if (isNaN(a) || isNaN(b)) {
        res.status(400).send({ "error": 'Uno de los parámetros no es un número' });
    } else {
        const result = core.add(a, b);

        await createHistoryEntry({ firstArg: a, secondArg: b, result, operationName: "ADD", error: null })
        return res.send({ result });
    }
});


router.get("/pow/:a", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);

    if (isNaN(a)) {
        res.status(400).send({ "error": 'Uno de los parámetros no es un número' });
    } else {
        const result = core.pow(a);

        await createHistoryEntry({ firstArg: a, secondArg: null, result, operationName: "POW", error: null })
        return res.send({ result });
    }
});


router.get("/div/:a/:b", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);
    const b = Number(params.b);

    if (isNaN(a) || isNaN(b)) {
        res.status(400).send({ "error": 'Uno de los parámetros no es un número' });
    } else {
        const result = core.div(a, b);
        if (result === "Math Error") {
            res.status(400).send({ "error": 'El divisor no puede ser cero' });
        } else {
            await createHistoryEntry({ firstArg: a, secondArg: b, result, operationName: "DIV", error: null })
            return res.send({ result });
        }
    }
});

router.get("/mul/:a/:b", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);
    const b = Number(params.b);

    if (isNaN(a) || isNaN(b)) {
        res.status(400).send({ "error": 'Uno de los parámetros no es un número' });
    } else {
        const result = core.mul(a, b);

        await createHistoryEntry({ firstArg: a, secondArg: b, result, operationName: "MUL", error: null })
        return res.send({ result });
    }
});

router.get("/sqrt/:a", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);

    if (isNaN(a)) {
        res.status(400).send({ "error": 'El parámetro no es un numero' });
    } else {
        const result = core.sqrt(a);

        await createHistoryEntry({ firstArg: a, result, operationName: "SQRT", error: null })
        return res.send({ result });
    }
});

router.get("/binary/:a", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);

    if (isNaN(a)) {
        res.status(400).send({ "error": 'El parámetro no es un numero' });
    } else {
        const result = core.binary(a);

        await createHistoryEntry({ firstArg: a, result, operationName: "BINARY", error: null })
        return res.send({ result });
    }
});

router.get("/all/:operation?", async function (req, res) {
    try {
        const { operation } = req.params;
        let result;

        if (operation === "mul" || operation === "div" || operation === "add" || operation === "sub" || operation === "pow" || operation === "sqrt" || operation === "binary") {
            result = await getHistory(operation.toUpperCase());
        } else if (!operation) {
            result = await getHistory();
        }
        else {
            return res.status(400).send({ "error": 'El parámetro no es una operacion' });
        }

        return res.send({ result });
    } catch (error) {
        return res.status(500).send({ error: 'Error al obtener el historial' });
    }
});

router.get("/allDeleted", async function (req, res) {
    try {
        const rowsDeleted = await deleteAllHistory();
        console.log(`${rowsDeleted} registros eliminados de la tabla History.`);
        res.status(200).send({ "result": `${rowsDeleted} registros eliminados de la tabla History.` });
    } catch (error) {
        console.error('Error al eliminar los registros de History:', error);
        res.status(500).send({ "result": 'Error al eliminar los registros de History.' });
    }
});

router.get("/id/:id", async function (req, res) {
    const id = req.params.id;

    try {
        const result = await getHistoryEntryById(id);

        if (!result) {
            return res.status(404).send({ error: 'No se encontró ninguna entrada de historial con el ID proporcionado' });
        }

        return res.send({ result });
    } catch (error) {
        console.error('Error al obtener la entrada del historial:', error);
        return res.status(500).send({ error: 'Error al obtener la entrada del historial' });
    }
});

export default router;